File: articles/0004_std_lib_http_client/src/python/tests/test_unix_transport.py
import pytest
import socket
import threading
import time
import os
from queue import Queue
from unittest.mock import patch

from httppy.unix_transport import UnixTransport
from httppy.errors import (
    SocketConnectError,
    SocketWriteError,
    TransportError,
    SocketReadError
)

class UnixServerFixture:
    def __init__(self, handler):
        self._handler = handler
        self._should_stop = threading.Event()

        self.socket_path = f"/tmp/httppy_test_{os.getpid()}_{time.time_ns()}"
        # Clean up any old socket file
        if os.path.exists(self.socket_path):
            os.remove(self.socket_path)

        self.listener_sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self.listener_sock.settimeout(0.5) # To prevent hanging on accept
        self.listener_sock.bind(self.socket_path)
        self.thread = threading.Thread(target=self._accept_loop)

    def start(self):
        self.thread.start()

    def stop(self):
        if not self._should_stop.is_set():
            self._should_stop.set()
            self.thread.join()
            self.listener_sock.close()
            if os.path.exists(self.socket_path):
                os.remove(self.socket_path)

    def _accept_loop(self):
        self.listener_sock.listen()
        while not self._should_stop.is_set():
            try:
                client_sock, _ = self.listener_sock.accept()
                with client_sock:
                    if self._handler:
                        self._handler(client_sock)
                # Only handle one connection for simplicity in tests
                break
            except socket.timeout:
                continue
            except OSError:
                break

@pytest.fixture
def test_server(request):
    server = UnixServerFixture(request.param if hasattr(request, "param") else None)
    server.start()
    yield server
    server.stop()


def test_construction_succeeds():
    try:
        _ = UnixTransport()
        assert True
    except Exception:
        pytest.fail("UnixTransport construction failed")

@pytest.mark.parametrize("test_server", [lambda sock: None], indirect=True)
def test_connect_succeeds(test_server):
    transport = UnixTransport()
    transport.connect(test_server.socket_path, 0)
    transport.close()

def test_write_succeeds(test_server):
    message_queue = Queue()
    def server_logic(sock):
        data = sock.recv(1024)
        message_queue.put(data)
    test_server._handler = server_logic

    transport = UnixTransport()
    transport.connect(test_server.socket_path, 0)
    message_to_send = b"hello from client"
    bytes_written = transport.write(message_to_send)

    assert bytes_written == len(message_to_send)
    captured_message = message_queue.get(timeout=1)
    assert captured_message == message_to_send
    transport.close()

@pytest.mark.parametrize("test_server", [lambda sock: sock.sendall(b"hello from server")], indirect=True)
def test_read_into_succeeds(test_server):
    transport = UnixTransport()
    transport.connect(test_server.socket_path, 0)

    buffer = bytearray(1024)
    bytes_read = transport.read_into(buffer)

    assert bytes_read == len(b"hello from server")
    assert buffer[:bytes_read] == b"hello from server"
    transport.close()

@pytest.mark.parametrize("test_server", [lambda sock: None], indirect=True)
def test_close_succeeds(test_server):
    transport = UnixTransport()
    transport.connect(test_server.socket_path, 0)
    transport.close()

@pytest.mark.parametrize("test_server", [lambda sock: None], indirect=True)
def test_close_is_idempotent(test_server):
    transport = UnixTransport()
    transport.connect(test_server.socket_path, 0)
    transport.close()
    transport.close()

def test_connect_fails_on_unresponsive_path():
    transport = UnixTransport()
    path = "/tmp/this-is-a-non-existent-socket.sock"
    with pytest.raises(SocketConnectError):
        transport.connect(path, 0)

def test_write_fails_on_closed_connection():
    server_closed_event = threading.Event()
    def server_logic(sock):
        import struct
        l_onoff = 1
        l_linger = 0
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_LINGER, struct.pack('ii', l_onoff, l_linger))
        server_closed_event.set()

    server = UnixServerFixture(server_logic)
    server.start()

    transport = UnixTransport()
    transport.connect(server.socket_path, 0)

    server_closed_event.wait(timeout=1)
    time.sleep(0.05)

    with pytest.raises(SocketWriteError):
        transport.write(b"this should fail")

    server.stop()

@pytest.mark.parametrize("test_server", [lambda sock: None], indirect=True)
def test_read_into_fails_on_peer_shutdown(test_server):
    transport = UnixTransport()
    transport.connect(test_server.socket_path, 0)

    test_server.thread.join(timeout=1)

    buffer = bytearray(1024)
    bytes_read = transport.read_into(buffer)
    assert bytes_read == 0
    transport.close()

def test_write_fails_if_not_connected():
    transport = UnixTransport()
    with pytest.raises(TransportError, match="Cannot write on a disconnected transport."):
        transport.write(b"some data")

def test_read_into_fails_if_not_connected():
    transport = UnixTransport()
    buffer = bytearray(1024)
    with pytest.raises(TransportError, match="Cannot read from a disconnected transport."):
        transport.read_into(buffer)

@pytest.mark.parametrize("test_server", [lambda sock: None], indirect=True)
def test_connect_fails_if_already_connected(test_server):
    transport = UnixTransport()
    transport.connect(test_server.socket_path, 0)

    with pytest.raises(TransportError, match="Transport is already connected."):
        transport.connect(test_server.socket_path, 0)

    transport.close()

@pytest.mark.parametrize("test_server", [lambda sock: None], indirect=True)
def test_read_into_raises_socket_read_error_on_os_error(test_server):
    transport = UnixTransport()
    transport.connect(test_server.socket_path, 0)

    buffer = bytearray(1024)

    with patch('socket.socket.recv_into') as mock_recv_into:
        mock_recv_into.side_effect = OSError("Mock OS-level read error")

        with pytest.raises(SocketReadError, match="Mock OS-level read error"):
            transport.read_into(buffer)

    transport.close()

File: articles/0004_std_lib_http_client/tests/cpp/test_unix_transport.cpp
#include <gtest/gtest.h>
#include <httpcpp/unix_transport.hpp>

#include <thread>
#include <atomic>
#include <string>
#include <vector>
#include <csignal>
#include <functional>
#include <future>

#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>

class UnixTransportTest : public ::testing::Test {
protected:
    std::thread server_thread_;
    std::atomic<bool> should_stop_{false};
    int listener_fd_ = -1;

    std::string socket_path_;
    std::string captured_message_;
    std::function<void(int)> server_logic_;

    httpcpp::UnixTransport transport_;

    void SetUp() override {
        signal(SIGPIPE, SIG_IGN);
    }

    void TearDown() override {
        StopServer();
    }

    void StartServer(std::function<void(int)> server_logic) {
        server_logic_ = std::move(server_logic);

        socket_path_ = "/tmp/httpcpp_test_" + std::to_string(getpid());
        unlink(socket_path_.c_str());

        listener_fd_ = socket(AF_UNIX, SOCK_STREAM, 0);
        ASSERT_NE(listener_fd_, -1);

        sockaddr_un serv_addr{};
        serv_addr.sun_family = AF_UNIX;
        strncpy(serv_addr.sun_path, socket_path_.c_str(), sizeof(serv_addr.sun_path) - 1);

        ASSERT_EQ(bind(listener_fd_, (struct sockaddr*)&serv_addr, sizeof(serv_addr)), 0);
        ASSERT_EQ(listen(listener_fd_, 1), 0);

        server_thread_ = std::thread(&UnixTransportTest::AcceptLoop, this);
    }

    void StopServer() {
        if (!should_stop_.exchange(true)) {
            if (listener_fd_ != -1) {
                shutdown(listener_fd_, SHUT_RDWR);
            }
            if (server_thread_.joinable()) {
                server_thread_.join();
            }
            if (listener_fd_ != -1) {
                close(listener_fd_);
                unlink(socket_path_.c_str());
            }
        }
    }

private:
    void AcceptLoop() {
        int client_fd = accept(listener_fd_, nullptr, nullptr);
        if (client_fd < 0) {
            return;
        }
        if (server_logic_) {
            server_logic_(client_fd);
        }
        close(client_fd);
    }
};

TEST(UnixTransportLifecycle, ConstructionSucceeds) {
    httpcpp::UnixTransport transport;
    SUCCEED();
}

TEST_F(UnixTransportTest, ConnectSucceeds) {
    StartServer([](int){});
    auto result = transport_.connect(socket_path_.c_str(), 0);
    ASSERT_TRUE(result.has_value());
}

TEST_F(UnixTransportTest, WriteSucceeds) {
    std::promise<void> server_read_promise;
    auto server_read_future = server_read_promise.get_future();

    StartServer([this, &server_read_promise](int client_fd){
        std::vector<char> buffer(1024, 0);
        ssize_t bytes_read = read(client_fd, buffer.data(), buffer.size() - 1);
        if (bytes_read > 0) {
            captured_message_.assign(buffer.data(), bytes_read);
        }
        server_read_promise.set_value();
    });

    auto connect_result = transport_.connect(socket_path_.c_str(), 0);
    ASSERT_TRUE(connect_result.has_value());

    const std::string message = "hello unix server";
    auto write_result = transport_.write({
        reinterpret_cast<const std::byte*>(message.data()), message.size()
    });
    ASSERT_TRUE(write_result.has_value());

    server_read_future.wait();
    ASSERT_EQ(captured_message_, message);
}

TEST_F(UnixTransportTest, ReadSucceeds) {
    const std::string server_message = "hello unix client";
    StartServer([&server_message](int client_fd){
        write(client_fd, server_message.c_str(), server_message.length());
    });

    auto connect_result = transport_.connect(socket_path_.c_str(), 0);
    ASSERT_TRUE(connect_result.has_value());

    std::vector<std::byte> buffer(1024);
    auto read_result = transport_.read(buffer);
    ASSERT_TRUE(read_result.has_value());
    ASSERT_EQ(*read_result, server_message.size());

    std::string received_message(reinterpret_cast<const char*>(buffer.data()), *read_result);
    ASSERT_EQ(received_message, server_message);
}

TEST_F(UnixTransportTest, CloseSucceeds) {
    StartServer([](int){});
    auto connect_result = transport_.connect(socket_path_.c_str(), 0);
    ASSERT_TRUE(connect_result.has_value());
    auto close_result = transport_.close();
    ASSERT_TRUE(close_result.has_value());
}

TEST_F(UnixTransportTest, CloseIsIdempotent) {
    StartServer([](int){});
    auto connect_result = transport_.connect(socket_path_.c_str(), 0);
    ASSERT_TRUE(connect_result.has_value());
    ASSERT_TRUE(transport_.close().has_value());
    ASSERT_TRUE(transport_.close().has_value());
}

TEST_F(UnixTransportTest, ConnectFailsOnUnresponsiveSocket) {
    auto result = transport_.connect("/tmp/non-existent-socket-path.sock", 0);
    ASSERT_FALSE(result.has_value());
    ASSERT_EQ(result.error(), httpcpp::TransportError::SocketConnectFailure);
}

TEST_F(UnixTransportTest, ReadFailsOnPeerShutdown) {
    std::promise<void> server_accepted_promise;
    auto server_accepted_future = server_accepted_promise.get_future();
    StartServer([&server_accepted_promise](int){
        server_accepted_promise.set_value();
    });

    auto connect_result = transport_.connect(socket_path_.c_str(), 0);
    ASSERT_TRUE(connect_result.has_value());
    server_accepted_future.wait();

    std::vector<std::byte> buffer(1024);
    auto read_result = transport_.read(buffer);
    ASSERT_FALSE(read_result.has_value());
    ASSERT_EQ(read_result.error(), httpcpp::TransportError::ConnectionClosed);
}

TEST_F(UnixTransportTest, WriteFailsIfNotConnected) {
    const std::string message = "test";
    auto write_result = transport_.write({
        reinterpret_cast<const std::byte*>(message.data()), message.size()
    });

    ASSERT_FALSE(write_result.has_value());
    ASSERT_EQ(write_result.error(), httpcpp::TransportError::SocketWriteFailure);
}

TEST_F(UnixTransportTest, ReadFailsIfNotConnected) {
    std::vector<std::byte> buffer(1024);
    auto read_result = transport_.read(buffer);

    ASSERT_FALSE(read_result.has_value());
    ASSERT_EQ(read_result.error(), httpcpp::TransportError::SocketReadFailure);
}

TEST_F(UnixTransportTest, ConnectFailsIfAlreadyConnected) {
    StartServer([](int){});

    auto connect_result1 = transport_.connect(socket_path_.c_str(), 0);
    ASSERT_TRUE(connect_result1.has_value());

    auto connect_result2 = transport_.connect(socket_path_.c_str(), 0);
    ASSERT_FALSE(connect_result2.has_value());
    ASSERT_EQ(connect_result2.error(), httpcpp::TransportError::SocketConnectFailure);
}

TEST_F(UnixTransportTest, WriteFailsOnClosedConnection) {
    std::promise<void> server_closed_promise;
    auto server_closed_future = server_closed_promise.get_future();

    StartServer([&server_closed_promise](int client_fd){
        struct linger so_linger{};
        so_linger.l_onoff = 1;
        so_linger.l_linger = 0;
        setsockopt(client_fd, SOL_SOCKET, SO_LINGER, &so_linger, sizeof(so_linger));
        server_closed_promise.set_value();
    });

    auto connect_result = transport_.connect(socket_path_.c_str(), 0);
    ASSERT_TRUE(connect_result.has_value());

    server_closed_future.wait();
    std::this_thread::sleep_for(std::chrono::milliseconds(50));

    const std::string message = "this will fail";
    auto write_result = transport_.write({
        reinterpret_cast<const std::byte*>(message.data()), message.size()
    });

    ASSERT_FALSE(write_result.has_value());
    ASSERT_EQ(write_result.error(), httpcpp::TransportError::SocketWriteFailure);
}